src/library/scala/concurrent/BatchingExecutor.scala

```diff
@@ -24,133 +24,124 @@ trait Batchable {
- * not use `scala.concurrent.blocking` when it should,
- * because tasks created within other tasks will block
- * on the outer task completing.
- * This executor may run tasks in any order, including LIFO order.
- * There are no ordering guarantees.
- *
- * WARNING: The underlying Executor's execute-method must not execute the submitted Runnable
- * in the calling thread synchronously. It must enqueue/handoff the Runnable.
- */
- private[concurrent] trait BatchingExecutor extends Executor {
-  private[this] final val _tasksLocal = new ThreadLocal[Batch]()
+private[concurrent] trait BatchingExecutor extends Executor {
+  private[this] final val _tasksLocal = new ThreadLocal[AnyRef]()
 
-  private[this] final class Batch extends Runnable with BlockContext with (BlockContext => Throwable) {
+  /*
+   * Batch implements a LIFO queue (stack) and is used as a trampolining Runnable.
+   * In order to conserve allocations, the first element in the batch is stored "unboxed" in
+   * the `first` field. Subsequent Runnables are stored in the array called `other`.
+  */
+  private[this] final class Batch(private[this] final val resubmitOnBlock: Boolean) extends Runnable with BlockContext with (BlockContext => Throwable) {
     private[this] final var parentBlockContext: BlockContext = _
     private[this] final var first: Runnable = _
     private[this] final var size: Int = _
     private[this] final var other: Array[Runnable] = BatchingExecutorStatics.emptyBatchArray
 
-    def this(r: Runnable) = {
-      this()
+    def this(r: Runnable, resubmitOnBlock: Boolean) = {
+      this(resubmitOnBlock)
       first = r
       size = 1
     }
 
-    private def this(first: Runnable, other: Array[Runnable], size: Int) = {
-      this()
+    private def this(first: Runnable, other: Array[Runnable], size: Int, resubmitOnBlock: Boolean) = {
+      this(resubmitOnBlock)
       this.first = first
       this.other = other
       this.size = size
     }
 
     private[this] final def cloneAndClear(): Batch = {
-      val newBatch = new Batch(first, other, size)
+      val newBatch = new Batch(first, other, size, resubmitOnBlock)
       this.first = null
       this.other = BatchingExecutorStatics.emptyBatchArray
       this.size = 0
       newBatch
     }
 
-    private[this] final def grow(): Unit = {
-      val len = other.length
-      other =
-        if (len == 0) new Array[Runnable](4)
-        else {
-          val newOther = new Array[Runnable](len << 1)
-          System.arraycopy(other, 0, newOther, 0, len)
-          newOther
-        }
+    private[this] final def ensureCapacity(curSize: Int): Array[Runnable] = {
+      val curOther = this.other
+      val curLen = curOther.length
+      if (curSize <= curLen) curOther
+      else {
+        val newLen = if (curLen == 0) 4 else curLen << 1
+
+        if (newLen <= curLen) throw new StackOverflowError("Space limit of asynchronous stack reached: " + curLen)
```

Comment:
A bit uncertain on this, on one hand we could throw a "normal" exception, on the other hand, we'll get an OOME if we are unable to allocate a large enough array—which is a fatal exception. I'm leaning towards this as an acceptable solution.

Reply:
The only way for this to occur is if `newLen` overflows to `Int.MinValue`. Would it make sense in that case to just set it to `Int.MaxValue`, since an `Array` cannot exceed that length anyway?

Reply:
it makes sense to me to throw a VMError, but SOE doesn't feel like quite the right one to me. `VirtualMachineError` doesn't appear to be abstract - what do you think of throwing that instead?
also, I'm wondering if the max size should go closer to the limit, rather than stopping at `Int.MaxValue / 2 + 1`

Reply:
unless I'm misunderstanding. is the batch only grown from within itself? in that case, SOE makes more sense to me, but not if you've just added too many `Runnable`s from anywhere